File: backend/physics_ai_tutor/schemas/question.py
from enum import StrEnum
from typing import Annotated

from pydantic import BaseModel, StringConstraints

QuestionText = Annotated[
    str, StringConstraints(strip_whitespace=True, min_length=1, max_length=1000)
]
AnswerText = Annotated[
    str, StringConstraints(strip_whitespace=True, min_length=1, max_length=5000)
]


class QuestionStatus(StrEnum):
    UNREVIEWED = "UNREVIEWED"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"


class QuestionSource(StrEnum):
    MANUAL = "MANUAL"
    AI_GENERATED = "AI_GENERATED"


class QuestionCreate(BaseModel):
    question: QuestionText
    answer: AnswerText


class QuestionUpdate(QuestionCreate):
    pass


class QuestionResponse(BaseModel):
    id: int
    question: str
    answer: str
    status: QuestionStatus
    source: QuestionSource

    model_config = {
        "from_attributes": True
    }


class QuestionListResponse(BaseModel):
    items: list[QuestionResponse]
    total: int
    page: int
    size: int


class QuestionImportResponse(BaseModel):
    created_count: int
    questions: list[QuestionResponse]

File: frontend/src/types/question.ts
export type QuestionStatus = "UNREVIEWED" | "APPROVED" | "REJECTED";
export type QuestionSource = "MANUAL" | "AI_GENERATED";

export type Question = {
    id: number;
    question: string;
    answer: string;
    status: QuestionStatus;
    source: QuestionSource;
    language: string;
    concepts: string[];
};

export type QuestionListResponse = {
    items: Question[];
    total: number;
    page: number;
    size: number;
};

export type CreateQuestionRequest = {
    question: string;
    answer: string;
};

export type UpdateQuestionRequest = {
    id: number;
    question: string;
    answer: string;
};

export type SimilarQuestion = {
    id: number;
    question: string;
    answer: string;
    distance: number;
};

export type FetchQuestionsParams = {
    page?: number;
    size?: number;
    keyword?: string;
    status?: QuestionStatus;
};

export type SearchQuestionsParams = {
    query: string;
    limit?: number;
};

export type QuestionReviewAction = "APPROVE" | "EDIT_APPROVE" | "REJECT";

export type QuestionReview = {
    id: number;
    question_id: number;
    action: QuestionReviewAction;
    before_question: string | null;
    before_answer: string | null;
    after_question: string | null;
    after_answer: string | null;
    reviewer_id: number;
    comment: string | null;
    created_at: string;
};

export type ReviewQuestionRequest = {
    action: QuestionReviewAction;
    question?: string;
    answer?: string;
    comment?: string;
};

export type QuestionImportResponse = {
    created_count: number;
    questions: Question[];
};

export type BulkDeleteResponse = {
    deleted_count: number;
    not_found_ids: number[];
};

export type BulkReviewAction = "APPROVE" | "REJECT";

export type BulkReviewResponse = {
    questions: Question[];
    not_found_ids: number[];
};

export type ConceptExtractionResult = {
    question_id: number;
    success: boolean;
    concepts: string[];
};

export type ConceptExtractionBatchResponse = {
    results: ConceptExtractionResult[];
};
